### PR TITLE
1637 SingleDatasetPerformance update to support multiple scorers

### PR DESCRIPTION
**Pull request by TheSolY**
SingleDatasetScalarPerformance renamed SingleDatasetPerformance 
Support for multiple scorers
Support for scorers per class
reduce_output implementation

### CODE REVIEW

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+__all__ = ['SingleDatasetPerformance']
+
+
+class SingleDatasetPerformance(SingleDatasetCheck, ReduceMixin):
+    """Calculate a performance metric as a scalar for a given model and a given dataset.
+
+    Parameters
+    ----------
+    alternative_scorers : Dict[str, Callable], default: None
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:32]: Would name this scorers (as in tabular performance)

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        Parameters
+        ----------
+        threshold: float
+            The threshold that the metrics result should be grater than.
+
+        metrics: List[str]
+            The names of the metrics from the check to apply the condition to. If None, runs on all of the metrics that
+            were calculated in the check.
+
+       class_mode: str, default: 'all'
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:90]: Would rephrase a bit, it is not clear that you can pass a class name.
Add options of average and weighted average score between classes

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                else:
+                    raise DeepchecksValueError(f'class_mode expected be one of the classes in the check results or any '
+                                               f'or all, recieved {class_mode}.')
+
+            if all(metrics_pass):
+                return ConditionResult(ConditionCategory.PASS, 'Passed for all of the mertics.')
+            else:
+                failed_metrics = ([a for a, b in zip(metrics, metrics_pass) if not b])
+                return ConditionResult(ConditionCategory.FAIL, f'Failed for metrics: {failed_metrics}')
+        return self.add_condition(f'Score is greater than {threshold}', condition)
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:124]: Extend this text (it appears on the output)

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    def reduce_output(self, check_result: CheckResult) -> Dict[str, float]:
+        """Return the values of the metrics for the test dataset in {metric: value} format."""
+        # Find metrics that were reduced over the classes and replace the Class Name with None
+        is_reduced_metrics = check_result.value.groupby('Metric')['Class Name'].nunique() == 1
+        reduced_metrics = is_reduced_metrics[is_reduced_metrics].index.to_list()
+        check_result.value.loc[check_result.value.Metric.apply(lambda x: x in reduced_metrics), 'Class Name'] = None
+
+        # Dict keys are format metric_class
+        metric_class = check_result.value.loc[:, ['Metric', 'Class Name']].aggregate(lambda x:
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:72]: How does metrics that are originally aggregated will look like? for example f1_micro

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+from deepchecks.core.condition import ConditionCategory
+from deepchecks.core.errors import DeepchecksValueError
+from deepchecks.vision import Batch, Context, SingleDatasetCheck
+from deepchecks.vision.metrics_utils.metrics import get_scorers_dict, metric_results_to_df
+
+__all__ = ['SingleDatasetPerformance']
+
+
+class SingleDatasetPerformance(SingleDatasetCheck, ReduceMixin):
+    """Calculate a performance metric as a scalar for a given model and a given dataset.
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:28]:
```suggestion
    """Calculate performance metrics of a given model on a given dataset.
```

File: tests/vision/checks/model_evaluation/single_dataset_performance_test.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+    assert_that(result.conditions_results[3], equal_condition_result(
+        is_pass=False,
+        category=ConditionCategory.ERROR,
+        details='Exception in condition: DeepchecksValueError: class_mode expected be one of the classes in the check '
+                'results or any or all, recieved a.',
+        name='Score is greater than 0.5'
+        ))
+
+def test_reduce_output(mnist_dataset_test, mock_trained_mnist, device):
```
Comment (Nadav-Barak) [inline tests/vision/checks/model_evaluation/single_dataset_performance_test.py:106]: Would add an additional test that receives scorers a list of str

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+**Structure:**
+
+* `What Is the Purpose of the Check? <#what-is-the-purpose-of-the-check>`__
+* `Generate data an model <#generate-data-and-model>`__
+* `Run the check <#run-the-check>`__
+* `Define a condition <#define-a-condition>`__
+
+What Is the Purpose of the Check?
+=================================
+This check returns the results from a dict of metrics, in the format metric name: ignite.Metric, calculated for the
```
Comment (Nadav-Barak) [inline docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:19]: Would add explanation What Is the Purpose of the Check and when should you use it (on a test set in model evaluation/ on a new batch of data )

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+given model dataset.
```
Comment (Nadav-Barak) [inline docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:20]: Add additional information on the required scorer type + 
what are the default scorers here (instead of below)

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+mnist_model = mnist.load_model()
+train_ds = mnist.load_dataset(train=True, object_type='VisionData')
+
+#%%
+# Run the check
+# -------------
+# We will run the check with the model defined above.
+#
+# The check will use the default classification metrics - `Precision
```
Comment (Nadav-Barak) [inline docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:41]: Would add here information on the scorers argument: 
"You can select which scorers to use by passing either a list or a dict of scorers to the check,
# the full list of possible scorers can be seen at ..."

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 4 above)
Comment (TheSolY) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:72]: @Nadav-Barak  They will have just 1 line instead of line per class

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 1 above)
Comment (TheSolY) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py]: @Nadav-Barak renamed

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 2 above)
Comment (TheSolY) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:88]: @Nadav-Barak it's clarified 2 lines below, We decided that the reduce over classes (e.g. average) is part of the metric and not a part of the condition.

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 3 above)
Comment (TheSolY) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:122]: @Nadav-Barak what info is missing there IYO?

File: tests/vision/checks/model_evaluation/single_dataset_performance_test.py (hunk 6 above)
Comment (TheSolY) [reply tests/vision/checks/model_evaluation/single_dataset_performance_test.py:106]: @Nadav-Barak this is managed by the function get scorers dict, and therefore the test belongs to the metrics module tests and not here

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 3 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:122]: would add for which metrics / all metrics
also add format_percent function around threshold

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py (hunk 8 above)
Comment (TheSolY) [reply docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:20]: @Nadav-Barak  IMO the defaults belong in the example on hoe to use with the defaults and not here. We might include a link to the metrics guide after it's complete.

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py (hunk 9 above)
Comment (TheSolY) [reply docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:41]: @Nadav-Barak We'll probably add a link to metrics guide to all the relevant plots after it's complete

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 2 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:88]: Maybe remove this option all together.
It is weird that we offer any (max)/ all(min) / single class but not a list of classes, average, weighted options

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+__all__ = ['SingleDatasetPerformance']
+
+
+class SingleDatasetPerformance(SingleDatasetCheck, ReduceMixin):
+    """Calculate performance metrics of a given model on a given dataset.
+
+    Parameters
+    ----------
+    scorers : Dict[str, Callable], default: None
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:32]:
```suggestion
    scorers : Union[List[str], Dict[str, Union[str, Callable]]], default: None
        List of scorers to use. If None, use default scorers.
        Scorers can be supplied as a list of scorer names or as a dictionary of names and functions.
```

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        An optional dictionary of scorer name to scorer functions.
+        If none given, using default scorers
+
+    """
+
+    def __init__(self, scorers: Dict[str, Metric] = None, **kwargs):
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:38]:
```suggestion
    def __init__(self, scorers: Union[List[str], Dict[str, Union[str, Callable]]] = None, **kwargs):
```

Comment (Nadav-Barak) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:38]: Question, do we want Metric or general Callable

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            scorer.update((prediction, label))
+
+    def compute(self, context: Context, dataset_kind: DatasetKind.TRAIN) -> CheckResult:
+        """Compute the metric result using the ignite metrics compute method and reduce to a scalar."""
+        results_dict = {}
+        for name, scorer in self.scorers.items():
+            result = scorer.compute()
+            results_dict[name] = [result] if isinstance(result, Number) else result
+        result_df = metric_results_to_df(results_dict, context.get_data_by_kind(dataset_kind))
+        return CheckResult(result_df)
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:60]:
```suggestion
        return CheckResult(result_df, header='Single Dataset Performance')
```

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+    def reduce_output(self, check_result: CheckResult) -> Dict[str, float]:
+        """Return the values of the metrics for the test dataset in {metric: value} format."""
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:63]:
```suggestion
       """Return the values of the metrics for the dataset provided in a {metric: value} format."""
```

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 11 above)
Comment (TheSolY) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:38]: @Nadav-Barak this test is for vision so the API should be Metric, unless there are other instructions @noamzbr

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py (hunk 9 above)
Comment (Nadav-Barak) [reply docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:42]: Would still add this sentence, for now link to metrics.py then afterward we could change the link to the metric guide

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+            were calculated in the check.
+
+        class_mode: str, default: 'all'
+            The decision rule over the classes, one of 'any', 'all', class name. If 'any', passes if at least one class
+            result is above the threshold, if 'all' passes if all of the class results are above the threshold,
+            class name, passes if the result for this specified class is above the thershold.
+        """
+
+        def condition(check_result, metrics=metrics):
+            if metrics is None:
```
Comment (Nadav-Barak) [inline deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:94]:
```suggestion
            check_result.value['Metric'].unique() if metrics is None else metrics```

Better not to use self or pycharm is angry

Comment (Nadav-Barak) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py:94]: Better not to use self or pycharm is angry

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py (hunk 9 above)
Comment (TheSolY) [reply docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:42]: metrics.py is internal so we shouldn't link to the user

File: deepchecks/vision/checks/model_evaluation/single_dataset_performance.py (hunk 10 above)
Comment (TheSolY) [reply deepchecks/vision/checks/model_evaluation/single_dataset_performance.py]: @Nadav-Barak fixed to Metric

**Review by ItayGabbay — APPROVED** [review]
